src/osd/OSD.cc

```diff
@@ -10206,43 +10206,77 @@ void OSD::maybe_override_max_osd_capacity_for_qos()
+      double default_iops = 0.0;
+
+      // Get the current osd iops capacity
+      double cur_iops = cct->_conf.get_val<double>(config_option);
+
+      // Get the default max iops capacity
+      auto val = cct->_conf.get_val_default(config_option);
+      if (!val.has_value()) {
+        derr << __func__ << " Unable to determine default value of "
+            << config_option << dendl;
+        // Cannot determine default iops. Force a run of the OSD benchmark.
+        force_run_benchmark = true;
+      } else {
+        // Default iops
+        default_iops = std::stod(val.value());
+      }
+
+      // Determine if we really need to run the osd benchmark
+      if (!force_run_benchmark && (default_iops != cur_iops)) {
+        dout(1) << __func__ << " force_run_benchmark: " << force_run_benchmark
+                << std::fixed << std::setprecision(2)
+                << " default_iops: " << default_iops
+                << " cur_iops: " << cur_iops
+                << ". Skip OSD benchmark test." << dendl;
+        return;
+      }
+    }
+
+    // Run osd bench: write 100 4MiB objects with blocksize 4KiB
     int64_t count = 12288000; // Count of bytes to write
     int64_t bsize = 4096;     // Block size
     int64_t osize = 4194304;  // Object size
     int64_t onum = 100;       // Count of objects to write
     double elapsed = 0.0;     // Time taken to complete the test
+    double iops = 0.0;
     stringstream ss;
     int ret = run_osd_bench_test(count, bsize, osize, onum, &elapsed, ss);
     if (ret != 0) {
       derr << __func__
            << " osd bench err: " << ret
            << " osd bench errstr: " << ss.str()
            << dendl;
-    } else {
-      double rate = count / elapsed;
-      double iops = rate / bsize;
-      dout(1) << __func__
-              << " osd bench result -"
-              << std::fixed << std::setprecision(3)
-              << " bandwidth (MiB/sec): " << rate / (1024 * 1024)
-              << " iops: " << iops
-              << " elapsed_sec: " << elapsed
-              << dendl;
-
-      // Override the appropriate config option
-      if (store_is_rotational) {
-        cct->_conf.set_val(
-          "osd_mclock_max_capacity_iops_hdd", std::to_string(iops));
-      } else {
-        cct->_conf.set_val(
-          "osd_mclock_max_capacity_iops_ssd", std::to_string(iops));
-      }
+      return;
+    }
 
-      // Override the max osd capacity for all shards
-      for (auto& shard : shards) {
-        shard->update_scheduler_config();
-      }
+    double rate = count / elapsed;
+    iops = rate / bsize;
+    dout(1) << __func__
+            << " osd bench result -"
+            << std::fixed << std::setprecision(3)
+            << " bandwidth (MiB/sec): " << rate / (1024 * 1024)
+            << " iops: " << iops
+            << " elapsed_sec: " << elapsed
+            << dendl;
+
+    // Persist iops to the MON store
+    mon_cmd_set_config(config_option, std::to_string(iops));
```

Comment:
check return value here, and log an error if persisting it fails? Unlikely to happen, but worth noting in the log

Reply:
The method mon_cmd_set_config() does log an error if the persisting fails. See https://github.com/ceph/ceph/pull/42133/files/4912a2a037786562b99d38d32468d15231042c75#diff-bd62068a5506fe0cde4498da3869fc8ef428bd168d4eed95dbb878d36d8b9650R10349.

Do you prefer the error detection and logging to be moved to the caller?

Reply:
@jdurgin As indicated in my previous comment, the inner method mon_cmd_set_config() does log an error if persisting fails. However, in the calling method I had not handled the failure. As a fall-back, the iops value should be updated using 
set_val() that updates the in-memory config "values" map (maintained by the ConfigValues object). I will push this in my 
next update.

Reply:
Pushed. Please verify.